oss_src/unity/python/sframe/cython/cy_variant.pyx

```diff
@@ -33,170 +36,593 @@ from .cy_unity cimport is_function_closure_info
+        elif tr_code == VAR_TR_LIST or tr_code == VAR_TR_TUPLE:
+            ret_as_fl[i].set_list(flex_list())
+
+            if tr_code == VAR_TR_LIST:
+                sub_is_flex_type = _var_set_listlike_internal(sub_vv, ret_as_fl[i].get_list_m(), <list>x, False)
+            else:
+                sub_is_flex_type = _var_set_listlike_internal(sub_vv, ret_as_fl[i].get_list_m(), <tuple>x, False)
+
+            if sub_is_flex_type:
+                check_list_to_vector_translation(ret_as_fl[i])
+                element_stored_in_flex_list = True
+            else:
+                if writing_to_flexible_types:
+                    __move_flex_list_to_variant_vector(ret_as_vv, ret_as_fl, i)
+                    writing_to_flexible_types = False
+
+                variant_set_variant_vector(ret_as_vv[i], sub_vv)
+
+        elif tr_code == VAR_TR_DICT:
+            ret_as_fl[i].set_dict(flex_dict())
+
+            sub_is_flex_type = _var_set_dict_internal(sub_vm, ret_as_fl[i].get_dict_m(), <dict>x, False)
+
+            if sub_is_flex_type:
+                element_stored_in_flex_list = True
+            else:
+                if writing_to_flexible_types:
+                    __move_flex_list_to_variant_vector(ret_as_vv, ret_as_fl, i)
+                    writing_to_flexible_types = False
+
+                variant_set_variant_map(ret_as_vv[i], sub_vm)
+
+        elif tr_code == VAR_TR_ATTEMPT_OTHER_FLEXIBLE_TYPE:
+            ret_as_fl[i] = _translate_to_flexible_type(x)
+            element_stored_in_flex_list = True
+        else:
+            if writing_to_flexible_types:
+                __move_flex_list_to_variant_vector(ret_as_vv, ret_as_fl, i)
+                writing_to_flexible_types = False
+
+            _convert_to_variant_type(ret_as_vv[i], x, tr_code)
+
+        # See if we need to move that element over to the variant type list
+        if element_stored_in_flex_list and not writing_to_flexible_types:
+            variant_set_flexible_type(ret_as_vv[i], ret_as_fl[i])
+
+
+    return writing_to_flexible_types
+
+############################################################
+# Translation routines for a dict
+
+cdef inline bint _var_set_dict_internal(variant_map_type& ret_as_vm, flex_dict& ret_as_fd, dict d, bint require_varmap) except *:
+    """
+    Attempts conversion of a dictionary to a flex_dict or
+    variant_map_type.
+
+    Returns True if the dict was converted to a flex_dict, in which
+    case the data is stored in ret_as_fd.  Otherwise, the data is
+    stored in ret_as_vm.
+
+    This function is useful to properly handle the recursions possible
+    with list and dict types.
+    """
+
+    # First, check if all the keys are string values.  If they are, we
+    # are doing a variant_map_type.  Otherwise, encode it as a
+    # flexible type.
+
+    cdef bint output_is_varmap = True
+
+    for k in d.iterkeys():
+        if type(k) is not str:
+            output_is_varmap = False
+            break
```

Comment:
yeah. Generally, if we can contain something in a flexible_type, we should: it is the simpler type.